ad/internal/winrmhelper/winrm_gmsa.go

```diff
@@ -0,0 +1,464 @@
+		HomePage:                      SanitiseTFInput(d, "home_page"),
+		ManagedPasswordIntervalInDays: d.Get("managed_password_interval_in_days").(int),
+		Name:                          SanitiseTFInput(d, "name"),
+
+		TrustedForDelegation: d.Get("trusted_for_delegation").(bool),
+	}
+
+	// delegate
+	del := []string{}
+	delegate := d.Get("principals_allowed_to_delegate_to_account").(*schema.Set)
+	for _, d := range delegate.List() {
+		if d == "" {
+			continue
+		}
+		del = append(del, d.(string))
+	}
+	if del != nil {
+		gmsa.PrincipalsAllowedToDelegateToAccount = del
+	}
+
+	// principal(s) allowed to retreieve password
+	pass := []string{}
+	passwords := d.Get("principals_allowed_to_retrieve_managed_password").(*schema.Set)
+	for _, p := range passwords.List() {
+		if p == "" {
+			continue
+		}
+		pass = append(pass, p.(string))
+	}
+
+	if pass != nil {
+		gmsa.PrincipalsAllowedToRetrieveManagedPassword = pass
+	}
+
+	// kerberos encryption types
+	krb := []string{}
+	krbenc := d.Get("kerberos_encryption_type").(*schema.Set)
+	for _, k := range krbenc.List() {
+		if k == "" {
+			continue
+		}
+		krb = append(krb, strings.ToUpper(k.(string)))
+	}
+
+	if krb != nil {
+		gmsa.KerberosEncryptionType = krb
+	}
+
+	return &gmsa
+}
+
+// GetGmsaFromHost returns a gmsa struct based on data
+// retrieved from the AD Domain Controller.
+func GetGmsaFromHost(client *winrm.Client, guid string, execLocally bool) (*Gmsa, error) {
+	cmd := fmt.Sprintf("Get-ADServiceAccount -identity %q -properties *", guid)
+	result, err := RunWinRMCommand(client, []string{cmd}, true, false, execLocally)
+	if err != nil {
+		return nil, err
+	}
+
+	if result.ExitCode != 0 {
+		log.Printf("[DEBUG] stderr: %s\nstdout: %s", result.StdErr, result.Stdout)
+		return nil, fmt.Errorf("command Get-ADgmsa exited with a non-zero exit code %d, stderr: %s", result.ExitCode, result.StdErr)
+	}
+
+	g, err := unmarshallGmsa([]byte(result.Stdout))
+
+	if err != nil {
+		return nil, fmt.Errorf("error while unmarshalling gmsa json document: %s", err)
+	}
+
+	return g, nil
+}
+
+// unmarshallGmsa unmarshalls the incoming byte array containing JSON
+// into a Gmsa structure and populates all fields based on the data
+// extracted.
+func unmarshallGmsa(input []byte) (*Gmsa, error) {
+	var gmsa Gmsa
+	err := json.Unmarshal(input, &gmsa)
+	if err != nil {
+		log.Printf("[DEBUG] Failed to unmarshall a ADGmsa json document with error %q, document was %s", err, string(input))
+		return nil, fmt.Errorf("failed while unmarshalling ADGmsa json document: %s", err)
+	}
+
+	commaIdx := strings.Index(gmsa.DistinguishedName, ",")
+	gmsa.Container = gmsa.DistinguishedName[commaIdx+1:]
+
+	if gmsa.ExpirationString != "" {
+		log.Printf("[DEBUG] unmarshall :: converting expiration date to proper format (current value: %s)", gmsa.ExpirationString)
+		var regdate = regexp.MustCompile(`^\/Date\((.+)\)\/$`)
+		// extract unixtime date
+		match := regdate.FindStringSubmatch(gmsa.ExpirationString)
```

Comment:
The code below will panic if the regexp fails to find a match. In  order to avoid that we must
check the length of `match` and if it's not greater than 1 then return a parsing error.

Reply:
Agreed, 

based on this [test](https://play.golang.org/p/I56gj7OVU0F) , we can consider that the regex is not matching if the length is equal to 0.

So, what about something like this ? :

```
// extract unixtime date
match := regdate.FindStringSubmatch(gmsa.ExpirationString)

if len(match) == 0 {
	return nil, fmt.Errorf("Failed to unmarshall a ADGmsa json, expiration date format is not matching regex.")
}
```

Reply:
Yup that's perfect!